src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -761,23 +794,30 @@ protected SCMRevision retrieve(@NonNull final String revision, @NonNull final Ta
             context.wantTags(true);
             context.withoutRefSpecs();
         }
+        if (fullHashMatch != null) {
+            //since this would have been skipped if this was a head or a tag we can just return whatever
+            return new GitRefSCMRevision(new GitRefSCMHead(fullHashMatch, fullHashMatches.iterator().next()), fullHashMatch);
+        }
         if (shortHashMatch != null) {
             // woot this seems unambiguous
             for (String name: shortNameMatches) {
                 if (name.startsWith(Constants.R_HEADS)) {
                     listener.getLogger().printf("Selected match: %s revision %s%n", name, shortHashMatch);
                     // WIN it's also a branch
-                    return new SCMRevisionImpl(new SCMHead(StringUtils.removeStart(name, Constants.R_HEADS)),
+                    return new GitBranchSCMRevision(new GitBranchSCMHead(StringUtils.removeStart(name, Constants.R_HEADS)),
                             shortHashMatch);
+                } else if (name.startsWith(Constants.R_HEADS)) {
```

Comment:
I believe this is unreachable code.  I'm surprised the Java compiler didn't detect this.

[Line 804](https://github.com/jenkinsci/git-plugin/pull/577/files#diff-33b2f30e4119d2fd0e0bb8653c0c58bdR804) is the same condition and when that condition is satisfied, the method returns.

Reply:
See my [pull request](https://github.com/rsandell/git-plugin/pull/1) which proposes to change that to Constants.R_TAGS and extends a test to reach that code.

Reply:
I haven't yet performed any interactive testing of the code change, just code review and running tests with breakpoints set in a debugger.  Several breakpoints in changes were not reached during my session, but I've not yet investigated further.

Reply:
Yea, this must be a copy paste error :/